src/main/java/org/elasticsearch/plugins/PluginManager.java

```diff
@@ -19,674 +19,557 @@
-                Files.delete(pluginFile);
-            } catch (Exception ex) {
-                log("Failed to delete plugin file" + pluginFile + " " + ex);
+            if ((plugins == null) || (plugins.isEmpty())) {
+                terminal.println("Plugin [%s] was not found. (run `plugin list` to get the list of installed plugins)", fqn);
+                return ExitStatus.IO_ERROR;
             }
-        }
 
-        if (FileSystemUtils.hasExtensions(extractLocation, ".java")) {
-            debug("Plugin installation assumed to be site plugin, but contains source code, aborting installation...");
-            try {
-                IOUtils.rm(extractLocation);
-            } catch(Exception ex) {
-                debug("Failed to remove site plugin from path " + extractLocation + " - " + ex.getMessage());
+            for (PluginDescriptor plugin : plugins) {
+                terminal.println(VERBOSE, "Found %s (%s)", plugin.name(), checkVersion(plugin.version()));
             }
-            throw new IllegalArgumentException("Plugin installation assumed to be site plugin, but contains source code, aborting installation.");
-        }
 
-        // It could potentially be a non explicit _site plugin
-        boolean potentialSitePlugin = true;
-        Path binFile = extractLocation.resolve("bin");
-        if (Files.isDirectory(binFile)) {
-            Path toLocation = pluginHandle.binDir(environment);
-            debug("Found bin, moving to " + toLocation.toAbsolutePath());
-            if (Files.exists(toLocation)) {
-                IOUtils.rm(toLocation);
-            }
-            try {
-                FileSystemUtils.move(binFile, toLocation);
-            } catch (IOException e) {
-                throw new IOException("Could not move [" + binFile + "] to [" + toLocation + "]", e);
-            }
-            if (Files.getFileStore(toLocation).supportsFileAttributeView(PosixFileAttributeView.class)) {
-                final Set<PosixFilePermission> perms = new HashSet<>();
-                perms.add(PosixFilePermission.OWNER_EXECUTE);
-                perms.add(PosixFilePermission.GROUP_EXECUTE);
-                perms.add(PosixFilePermission.OTHERS_EXECUTE);
-                Files.walkFileTree(toLocation, new SimpleFileVisitor<Path>() {
-                    @Override
-                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-                        if (attrs.isRegularFile()) {
-                            Files.setPosixFilePermissions(file, perms);
+            if (terminal.verbosity().enabled(NORMAL)) {
+                boolean required = false;
+
+                // Iterates over all installed plugins to check if the plugin to uninstall is required by another plugin
+                for (PluginDescriptor plugin : local.list()) {
+                    Set<PluginDescriptor> dependencies = plugin.dependencies();
+                    if (dependencies != null) {
```

Comment:
I think we don't need to check for `null` here. We are using `dependencies.build()` to create this set and it can't generate `null`.